wordpress-mobile/WordPress-Android
# Don't rely on `sleep` to load images while generating screenshots

First off, this is basically a rewrite of the screenshot generation code. The original tests used very verbose assertions that were a bit brittle in terms of how they looked up specific elements. While there may be a benefit to that I was unaware of, as I was working through the task, I realized I was repeating myself a lot and extracted the code I was writing to a new set of methods. Once I started working this way, it seemed a lot cleaner, so I applied it to the rest of the testing code.

One other benefit to extracting these methods from the screenshot generation suite is that it'll be possible for those methods to be used for other UI testing down the road, which will hopefully benefit other developers, too!

Let me know if this should be broken down into smaller pieces or adjusted in any way!

**A few issues this PR resolves**:

- If the device size is unusual, the tests don't crash while logging out.
- There's no way for notification avatars not to show up in screenshots.
- The test suite now waits until all images have loaded, rather than a specific amount of time.
- Screenshots won't sometimes show the RecyclerView loading indicator in the reader view.
- Navigating back to the main activity is now more reliable across different device sizes.
- If the device is already logged in when the test begins, the wPLogout method will be run before beginning the test so that it doesn't fail.

**To test:**

- Run the screenshot generation suite on a machine that's been used to generate the production screenshots before and ensure it works as intended.
- Ensure that the stats screenshot is generated correctly – my test blog has a pitiful amount of traffic, so I haven't been able to test this as thoroughly as I would've liked.

**To review:**

- The implementation of how we track the number of images remaining for download in Aztec. There were many ways to accomplish this, including:
	- Modifying Aztec directly to allow access to this information. (Rejected because that seemed like an if-all-else-fails solution).
	- Creating a subclass of AztecImageLoader in the screenshot generation suite and injecting it into EditPostActivity. (Rejected because it would mean adding AztecImageLoader to the DI system, which can get complicated because of the separate testing application).
	- Patching AztecImageLoader to track images being downloaded, and EditPostActivity to allow access to the image loader instance. (Went with this because it was the smallest set of changes and seemed like the safest option to ensure no breaking changes).
